crates/environ/src/cranelift.rs

```diff
@@ -240,12 +319,45 @@ fn compile(env: CompileEnv<'_>) -> Result<ModuleCacheDataTupleType, CompileError
                 context.func.collect_debug_info();
             }
 
+            let mut func_env = FuncEnvironment::new(isa.frontend_config(), env.local, env.tunables);
+
+            // We use these as constant offsets below in
+            // `stack_limit_from_arguments`, so assert their values here. This
+            // allows the closure below to get coerced to a function pointer, as
+            // needed by `ir::Function`.
+            //
+            // Otherwise our stack limit is specially calculated from the vmctx
+            // argument, where we need to load the `*const VMInterrupts`
+            // pointer, and then from that pointer we need to load the stack
+            // limit itself. Note that manual register allocation is needed here
+            // too due to how late in the process this codegen happens.
+            //
+            // For more information about interrupts and stack checks, see the
+            // top of this file.
+            assert_eq!(func_env.offsets.vmctx_interrupts(), 0);
+            assert_eq!(func_env.offsets.vminterrupts_stack_limit(), 0);
+            context.func.stack_limit_from_arguments = Some(|pos, scratch| {
```

Comment:
What @bjorn3 was hinting at is that all input cranelift accepts (not cranelift-wasm, but cranelift itself) is currently representable in the the clif text format. This is really nice for testing, reproducibility, and test case reduction. For example, see "Don't Lose Our Copy-and-Paste-This-IR-and-it-is-a-Test Property" in https://github.com/bytecodealliance/wasmtime/issues/1146#issuecomment-566790342

By providing a closure that is called in the middle of cranelift compilation, and without a text representation, we're losing those nice testing properties.

In this specific case, I'm not sure what alternative is better here. This isn't really generic behavior that makes sense inside cranelift. And we don't have anything like "always inline functions/macros" right now (maybe https://github.com/bytecodealliance/wasmtime/pull/1267 could eventually be that, its motivation was similar, but for GC barriers).

@sunfishcode felt pretty strongly about the everything-is-in-clif-text property, so tagging him for ideas here.

Reply:
Would it be possible to add `translate_func_prologue` to `cranelift_wasm::FuncEnviron`, where we already have embedder hooks and can side step these issues? I think this would also make it so we wouldn't need to regalloc by hand here.

Reply:
I'm not really wed to this particular solution at all, and I agree it's a great property to have everything reflected in the text format. That being said though we need *some* solution here which actually works and accounts for the wasmtime-specific behavior of where the stack limit is located. I don't know of a great way to do this, but would be totally onboard with alternative ways.

I don't know how to make `translate_func_prologue` work since the stack limit check happens at such an odd time (machine code generation) not at cranelift IR generation time (when wasm is translated). Would this need some sort of magical sequence of instructions at the prologue which gets pattern-matched for the stack check?

Reply:
> I don't know how to make `translate_func_prologue` work since the stack limit check happens at such an odd time (machine code generation) not at cranelift IR generation time (when wasm is translated). Would this need some sort of magical sequence of instructions at the prologue which gets pattern-matched for the stack check?

I think we would need an instruction to get the current stack frame's size or something. Not sure how well this would work out.

Reply:
The impression I get though is that representing this in IR is the wrong idea because if we generate IR to do the stack check, then the actual stack adjustment happens in the function's prologue, which is too late then to perform the stack check. This is very machine-specific because it's part of the prologue/epilogue generation at a very particular part of codegen, so afaik it can't be very easily represented in IR without a lot of "magic", but others may know better routes!

Reply:
> I don't know how to make translate_func_prologue work since the stack limit check happens at such an odd time (machine code generation) not at cranelift IR generation time (when wasm is translated). Would this need some sort of magical sequence of instructions at the prologue which gets pattern-matched for the stack check?

For what it's worth, the way it's dealt with in Spidermonkey is that:

1. Spidermonkey has its own calling convention "baldrdash", used in Cranelift.
2. Cranelift won't generate the prologue and epilogue if the calling convention is baldrdash; Spidermonkey will.
3. Then the interrupt check can be done in Cranelift IR from the Spidermonkey side.
4. In addition to this, Spidermonkey has to use the `fallthrough_return` instead of a plain return, since it generates its own epilogue.

I imagine replicating this behavior for Wasmtime would be quite a hassle, so this might not be the best path here...

Reply:
Keeping `ir::Function` serializable makes it easier for us to extract standalone Cranelift testcases that don't require linking with surrounding applications. And in the future, we also hope to use this property to enable more compilation caching and remote/ipc compilation.

Global values are an expression language which you can use to define code templates that Cranelift can expand into actual code later (https://github.com/bytecodealliance/wasmtime/pull/1267 is a further step toward generalizing the concept and actually renames it to "template"). Currently, you can only do a few simple things with them: [load, iadd_imm, vmctx addresses, and symbolic addresses](https://docs.rs/cranelift-codegen/0.62.0/cranelift_codegen/ir/enum.GlobalValueData.html). That looks like it may actually be enough for what you need to do here, though we can add more operations if needed.

So the idea is, you create them, one "instruction" at a time, with [declare_global_value](https://docs.rs/cranelift-codegen/0.62.0/cranelift_codegen/ir/function/struct.Function.html#method.create_global_value) which returns the index of the global value which you can then use however you want within the `ir::Function`.

Reply:
@sunfishcode oh I think that's what @bjorn3 [recommended above](https://github.com/bytecodealliance/wasmtime/pull/1490#discussion_r410474016) a few minutes ago as well, but do you know how to work around the legalizer issue? Expanding a global value seems to require the legalizer to run, but since this is in prologue/epilogue generation that's too late for the legalizer to run it seems.